### PR TITLE
Expose C99/C++11 functions in printers

### LINKED ISSUE #11353: ccode to output expm1, log1p, exp2, log2
Since C99, and C++11 respectively, there is support for:
- [expm1](http://en.cppreference.com/w/c/numeric/math/expm1)
- [log1p](http://en.cppreference.com/w/c/numeric/math/log1p)
- [exp2](http://en.cppreference.com/w/c/numeric/math/exp2)
- [log2](http://en.cppreference.com/w/c/numeric/math/log2)

These offer better precision/performance. Hence it would be nice if we could teach `ccode` how to output these, perhaps enabled using a flag not to break backward-compatibility?

### LINKED ISSUE #11953: proposal: ccode of Max as concatenation of 2-parameter std::max
In  reference to the discussion with @bjodah in https://github.com/sympy/sympy/pull/11933:
ccode of `Max(...)` now simply returns `Max(...)` which is not a standard C function.
`std::max` only accepts 2 arguments, while sympy's Max accepts an arbitrary number of arguments.
Concatenation of n-1 two-argument calls of std:max could be a solution:
  `ccode(Max(x1,x2,x3,x4,x5))`
would then return
  `std::max(std::max(std::max(std::max(x1,x2),x3),x4),x5)`
Modern C++ compilers should generate quite optimal code in this case. Anyway, in practise, `Max` is very often used with just two arguments, in which case everything falls back to what's intuitively expected.
Typically, `Max` is used as a safe-keeping for other functions to receive their arguments in a valid range.
  `log(Max(x,0.000001))`
If there is only 1 argument, we just get the identity:
  `ccode(Max(x1))`
would just return
  `x1`
Not sure what 
  `ccode(Max(x1,x2,x3,x4,x5), user_functions={"Max":"my_max"})`
should do. Users targetting restraint hardware probably would also prefer two-argument versions.

Of course, the same proposal also addresses `Min`.

### LINKED ISSUE #11969: printing.ccode regression for Max()
The following works in version 1.0 but has since broken.
```
x = sp.symbols("x")
exp = sp.Max(0, x)
constants, not_c, c_expr = sp.printing.ccode(exp, human=False)
assert not not_c, not_c
```

### PR DESCRIPTION
**Pull request by bjodah**
### Summary
- New modules: `sympy.printing.cxxcode`, `sympy.codegen.cfunctions`
- New printers: `C99CodePrinter`, `CXX98CodePrinter`, `CXX11CodePrinter` & provisional `_CXX17CodePrinter`
- New kwarg in `ccode`: `standard`
- New function: `cxxcode`
- New SymPy classes for C99/C++11 functions (e.g. `expm1`, `log1p` etc.) (in `.codegen.cfunctions`).
- Fixes #11353 
- Fixes #11953
- Fixes #11969 

#### TODO
- [x] handle `standard` kwarg in `ccode`.
- [x] move `cfunctions` to `.util.codegen`
- [x] write docstrings and update sphinx docs.
- [x] Add a `C89CodePrinter` class and make `CCodePrinter` inherit and extend it while being marked as deprecated.

Eventually we would like to have a helper function along the lines of:

```
>>> smart_ccode(2**x + log(x)/log(2))
'exp2(x) + log2(x)'
>>> smart_ccode((2*exp(x) - 2)*(3*exp(y) - 3), typically={x: And(-.1 < x, x < .1)})
'6*expm1(x)*(exp(y) - 1)'
```

but there is already a lot of things to discuss around this PR so I thought this was an appropriate time for a review.

Some bullet points for discussion:
- how should we handle different standards? e.g. C89 vs. C99 vs. C11, e.g. now I created a subclass printer and set ~~`language='C99'`~~ `standard='C99'`
- The helper functions in `printing.cfunctions` now define `_eval_expand_func`, `_eval_rewrite_as_tractable` and `fdiff`, anything more that would be useful here?
- I introduced `CXX98CodePrinter` and `CXX11CodePrinter` in `printing.cxxcode`, using qualified names (`std::` whenever possible). ~~There is most certainly still work to be done here.~~
- I would have liked to introduce a `C89CodePrinter` (which would have been the `CCodePrinter` minus some C99 features which have crept in), but I think that it would confuse more than anything else: people will wonder what role `CCodePrinter` has then.

### CODE REVIEW

**Comment by moorepants** [comment]
If CCodePrinter has all of the syntax shared between versions of C and C++ would it by default be C89? Are recent versions a superset of C89 or are there API changes?

**Comment by bjodah** [comment]
I would say that later standards are supersets, one exception that comes to mind is `auto` changing meaning in C++11, but currently it is not used by the printers.

**Comment by moorepants** [comment]
If that is the case it seems fine to use CCodePrinter as a base class and create subclasses for each version and variation of the language. If the user uses CCodePrinter directly, they'll just get C89.

**Comment by asmeurer** [comment]
Maybe the flag should be called `standard`, to match `fcode`.

**Comment by bjodah** [comment]
@jhncls I added support for Min and Max to this PR, if you have the time, I'd be interested in your input on this.

**Review by asmeurer — COMMENTED** [review]
I think it would be better to put the functions in the codegen module.

**Comment by bjodah** [comment]
The current version of the functions in `cfunctions` cannot be used as keys in dictionaries
I just added a failing test to demonstrate:
```
File ".../sympy/sympy/printing/tests/test_ccode.py", line 567, in test_get_math_macros
    assert macros[1/Sqrt(2)] == 'M_SQRT1_2'
KeyError: 1/Sqrt(2)
```
I would need to make the classes immutable, can we do so easily in SymPy? (I am only vaguely familiar with `Basic` and the metaclasses etc. that I think goes into that).

Or should I just avoid using such keys in dictionaries?

**Comment by asmeurer** [comment]
Your test fails with ImportError. Hashing should work just fine since you've subclassed from Function.

**Comment by bjodah** [comment]
Got it, thanks. (spurious `.pyc` file - should do `git clean -xfd` more often).

**Comment by bjodah** [comment]
@tommyolofsson - this should fix Min/Max printing, any feedback you may have is much appreciated.

**Comment by tommyolofsson** [comment]
@bjodah Looking good! I really appreciate that you kept C89 compatibility as well.

**Comment by asmeurer** [comment]
Maybe CCodePrinter should be an alias for C89CodePrinter. That will keep compatibility, but make it clear what standard it uses (and that other standards are available).

**Comment by asmeurer** [comment]
Maybe add some docstrings/doctests to the cfunctions (and add them to the Sphinx docs).

**Comment by bjodah** [comment]
Using `CCodePrinter` as an alias for `C89CodePrinter` would make sense, however the current version of `CCodePrinter` actually uses some C99 features. We would need to remove those then, but that would be a backward incompatible change.

**Comment by bjodah** [comment]
@jhncls yes of course (missed that when copy-pasting). thanks!

**Comment by asmeurer** [comment]
So maybe we should keep it as an alias, but deprecate it. Something like

```py
@deprecated(...)
class CCodePrinter(C89CodePrinter):
    """
    Deprecated.

    Alias for C89CodePrinter, for backwards compatibility.
    """
```

**Comment by bjodah** [comment]
I've started working on splitting the `CCodePrinter`, I've also started introducing a C89CodeGen and a C99CodeGen class, tests aren't passing for the codegen part but I've pushed my changes since there was interest in this work in #12220.

**Comment by bjodah** [comment]
@asmeurer I think this is ready for review. I've deprecated `CCodePrinter` and updated the codegen module accordingly.

**Comment by asmeurer** [comment]
I'm not sure that I like that ccode always raises a deprecation warning now. Are we going to make the standard argument required? 

This is our policy, by the way https://github.com/sympy/sympy/wiki/Deprecating-policy.

File: sympy/printing/ccode.py [hunk 1]
```diff
@@ -294,7 +321,107 @@ def indent_code(self, code):
         return pretty
 
 
-def ccode(expr, assign_to=None, **settings):
+class _C9XCodePrinter(object):
+    # Move these methods to C99CodePrinter when removing CCodePrinter
+    def _get_loop_opening_ending(self, indices):
+        open_lines = []
+        close_lines = []
+        loopstart = "for (int %(var)s=%(start)s; %(var)s<%(end)s; %(var)s++){"  # C99
```
Comment (asmeurer) [inline sympy/printing/ccode.py]: Is this the only incompatibility between the current CCodePrinter and C89Printer? This doesn't seem like much, not particularly worth making `ccode` always give a deprecation warning.

**Comment by asmeurer** [comment]
I'm only worried about the deprecation warnings from ccode, which seem like they could be quite annoying. The rest looks fine. I admit I'm not very proficient in C++, so hopefully I haven't missed anything obvious, but I think if there are any mistakes here we won't catch them until we merge this and people start using it.

File: sympy/printing/ccode.py (hunk 1 above)
Comment (bjodah) [inline sympy/printing/ccode.py]: That's the only user visible change I think. `asinh`, `acosh`, `atanh` & `tgamma` are also C99 functions. But I haven't added print methods for those so they are still supported by the C89 codeprinter even though they shouldn't.

**Comment by bjodah** [comment]
Regarding deprecation, I can definitely change back to not raising any deprecation warnings.
We could even make `'c99'` the default for the `standard` kwarg of `ccode` and thereby require users to only specify standard if they want something else.

Regarding the C++ code, I think so too. I've used this branch myself for some things but I can't really say that I've tested it exhaustively. One alternative is that I mark the API as "provisional" for a version or two, i.e. we reserve the right to change it without a deprecation cycle?

**Comment by asmeurer** [comment]
The specific output of the code printers does not count as API in my opinion. So we just have the new function `cxxcode` (which should probably be added to `__init__.py`), the codegen functions, and the code printer classes, which can be subclassed. Are there other things?

I don't know what the default standard should be. The advantage of c99 is that people who just use the function without realizing there is a standard option will see that it works for certain things that wouldn't work in c89. It does potentially break stuff for people using the old standard, but I don't see it as that major. I noticed that fcode defaults to 77, but perhaps C is more likely to be using the newer standards than Fortran or C++ (do you agree?).

**Comment by bjodah** [comment]
I think those are the things (can't think of anything else).

It's hard to say what people in general use, I know that there has been [discussions](https://gcc.gnu.org/ml/gcc/2014-10/msg00053.html) about using C11 as default in gcc. SymPy is a relatively young project. We should definitely support generating C89 code for e.g. people targeting embedded systems, but I think those user are fine with passing `stadnard='c89'` as a kwarg.

Should I try to remove the deprecation of `CCodePrinter` and just let it be an alias for `C99CodePrinter`? (and make the additions to `__init__.py` that you pointed out)

**Comment by asmeurer** [comment]
We can keep the deprecation, but make C99CodePrinter the default in ccode. That way ccode doesn't raise a warning, but anyone subclassing from CCodePrinter will still see it. That was what I had originally meant by deprecating it, so that we can remove the class in the future.

**Comment by bjodah** [comment]
@asmeurer so now there's considerably less deprecation warnings being raised. Good to merge?

File: sympy/printing/tests/test_ccode.py [hunk 2]
```diff
@@ -442,7 +449,9 @@ def test_ccode_loops_multiple_terms():
 
 def test_dereference_printing():
     expr = x + y + sin(z) + z
-    assert ccode(expr, dereference=[z]) == "x + y + (*z) + sin((*z))"
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=SymPyDeprecationWarning)
+        assert ccode(expr, dereference=[z]) == "x + y + (*z) + sin((*z))"
```
Comment (asmeurer) [inline sympy/printing/tests/test_ccode.py]: Why does this one still raise a warning?

**Comment by asmeurer** [comment]
That looks a lot better.

File: sympy/printing/tests/test_ccode.py (hunk 2 above)
Comment (bjodah) [inline sympy/printing/tests/test_ccode.py]: That was an oversight, I fixed it in the last commit.

**Comment by asmeurer** [comment]
Can you add something to the release notes for this https://github.com/sympy/sympy/wiki/Release-Notes-for-1.1

**Comment by bjodah** [comment]
Thanks, I've updated the release notes.

**Pull request merged**